ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A reference simulator of the RealPBFT protocol under \"ideal
+-- circumstances\"
+--
+module Test.Dynamic.Ref.RealPBFT (
+  Outcome (..),
+  State (..),
+  advanceUpTo,
+  emptyState,
+  nextSlot,
+  nullState,
+  step,
+  viable,
+  ) where
+
+import           Data.Foldable (Foldable, foldl', toList)
+import qualified Data.Map as Map
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+
+import           Ouroboros.Network.Block (SlotNo (..))
+
+import           Ouroboros.Consensus.NodeId (CoreNodeId (..))
+import           Ouroboros.Consensus.Protocol.Abstract (SecurityParam (..))
+import           Ouroboros.Consensus.Protocol.PBFT (PBftParams (..))
+
+import           Test.Dynamic.Util.NodeJoinPlan
+
+oneK :: Num a => PBftParams -> a
+oneK PBftParams{pbftSecurityParam} =
+    fromIntegral (maxRollbacks pbftSecurityParam)
+
+twoK :: Num a => PBftParams -> a
+twoK PBftParams{pbftSecurityParam} =
+    2 * fromIntegral (maxRollbacks pbftSecurityParam)
+
+oneN :: Num a => PBftParams -> a
+oneN PBftParams{pbftNumNodes} = fromIntegral pbftNumNodes
+
+{-------------------------------------------------------------------------------
+  PBFT state
+-------------------------------------------------------------------------------}
+
+-- | Outcome of a node leading a slot
+--
+data Outcome
+
+  = Absent
+    -- ^ the leader hadn't yet joined
+
+  | Nominal
+    -- ^ the leader extended the networks' common chain with a valid block
+
+  | Unable
+    -- ^ the leader could not forge a valid block
+
+  | Wasted
+    -- ^ the leader forged a valid-but-irrelevant block, because it forged
+    -- before sufficiently synchronizing
+    --
+    -- As of commit 4eb23cfe \"Merge #1260\", this only happens if a node leads
+    -- in the slot it joins. In that slot, it forges before the mini protocols
+    -- are able to pull in any of the existing nodes' chain, and so it
+    -- re-forges the epoch 0 boundary block (EBB).
+
+  deriving (Eq, Show)
+
+-- | Sliding windows and the next slot
+--
+-- The first window is who forged each of the last @k@ blocks, the second
+-- window is the outcomes of the last @2k@ slots. We also incrementally
+-- maintain the count of 'Nominal' outcomes in the window.
+--
+data State = State !(Seq CoreNodeId) !(Seq Outcome) !NumNominals !SlotNo
```

Comment:
Why not use a record?

Reply:
Fixed, thanks.

Reply:
(Just for the record, I hesitated to make it a record since: the field types were unique, I preferred a particular field order, and alphabetically field names wouldn't have preserved that order. But it was a weak preference.)